keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,355 @@
+- EndpointSlice will be covered by resource quota. This is to limit the max number of EndpointSlice objects in one namespace. This would provide protection for k8s apiserver. For instance, a malicious user would not be able to DOS k8s API by creating services selecting all pods.
+
+### Multiple IPs
+Each endpoint can contain more than one IP corresponding to the pod. The API itself does not enforce the consumer's behavior (e.g. kube-proxy).
+
+### Optional Endpoint Port
+The endpoint port number becomes optional in the EndpointSlice API while the port number field in core/v1 Endpoints API is required. This allows the API to support services with no port remapping or all port services.   
+
+### EndpointSlice Naming
+Use generateName with service name as prefix:
+```
+${service name}.${random}
+```
+
+### Label
+For all EndpointSlice objects managed by EndpointSlice controller. The following label is added to identify corresponding service:
+
+- Key: kubernetes.io/service
+- Value: ${service name}
+
+For EndpointSlice instances that are not derived from kubernetes Services, this label must not be applied.
+
+## Estimation
+This section provides comparisons between Endpoints API and EndpointSlice API under 3 scenarios:
+- Service Creation/Deletion
+- Single Endpoint Update
+- Rolling Update
+
+
+```
+Number of Backend Pod: P
+Number of Node: N
+Number of Endpoint Per EndpointSlice:B 
+```
+
+## Sample Case 1: 20,000 endpoints, 5,000 nodes
+ 
+### Service Creation/Deletion
+
+
+|                          | Endpoints             | 100 Endpoints per EndpointSlice | 1 Endpoint per EndpointSlice |
+|--------------------------|-----------------------|---------------------------------|------------------------------|
+| # of writes              | O(1)                  | O(P/B)                          | O(P)                         |
+|                          | 1                     | 200                             | 20000                        |
+| Size of API object       | O(P)                  | O(B)                            | O(1)                         |
+|                          | 20k * const = ~2.0 MB | 100 * const = ~10 KB            | < ~1KB                       |
+| # of watchers per object | O(N)                  | O(N)                            | O(N)                         |
+|                          | 5000                  | 5000                            | 5000                         |
+| # of total watch event   | O(N)                  | O(NP/B)                         | O(NP)                        |
+|                          | 5000                  | 5000 * 200 = 1,000,000          | 5000 * 20000 = 100,000,000   |
+| Total Bytes Transmitted  | O(PN)                 | O(PN)                           | O(PN)                        |
+|                          | 2.0MB * 5000 = 10GB   | 10KB * 5000 * 200 = 10GB        | ~10GB                        |
+
+### Single Endpoint Update
```

Comment:
See my last comment-- I think it would be better to compare not a single update, but rather a *rate* of updates, since that is what will happen in practice. In practice the rate will likely depend on the quality of the pods, multiplied by the size of the service--as services get bigger, holding the quality of the pods constant, the rate will go up.

Considering this, I think some entries that are currently labeled O(N) actually in practice are going to represent O(N * rate) = O(N^2) total load on the system; granted, the constant factor will be favorable.

Reply:
Under the same scenario (e.g. size of the service), can we assume the rate of change for both APIs is the same? Since we could not determine the rate of changes on pods, both API will respond to them. The main difference is that for each change, the cost is smaller.

Reply:
Just read you response down below. I think I understand your point now. I will respond.